### PR TITLE
feat: Add invocation ID and attempt count in x-goog-api-client header

### LINKED ISSUE #8881: storage: Add invocation-id to x-goog-api-client header for c# storage
Tracking ticket to add a unique invocation-id to the x-goog-api-client header for all Storage API requests.
See [nodejs  example](https://github.com/googleapis/nodejs-storage/pull/1920)

### PR DESCRIPTION
**Pull request by jskeet**
Fixes #8881

### CODE REVIEW

**Comment by jskeet** [comment]
Note: the behavior of the attempt count is still being checked, hence why this is a draft. It's unclear whether it's meant to start at "1" for the first request, or "1" for the *second* request (i.e. counting previous requests).

**Comment by jskeet** [comment]
(I'd expected this to be more complex and require GAX work, but it's quite simple in the end.)

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/RetryHandler.cs [hunk 1]
```diff
@@ -37,12 +46,17 @@ internal sealed class RetryHandler : IHttpUnsuccessfulResponseHandler
...
         };
-        private static RetryHandler s_instance = new RetryHandler();
+        private static readonly RetryHandler s_instance = new RetryHandler();
 
         private RetryHandler() { }
 
-        internal static void MarkAsRetriable<TResponse>(StorageBaseServiceRequest<TResponse> request) =>
+        internal static void MarkAsRetriable<TResponse>(StorageBaseServiceRequest<TResponse> request)
+        {
+            // Note: we can't use ModifyRequest, as the header is added later by ConfigurableMessageHandler.
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/RetryHandler.cs:55]:
```suggestion
            // Note: we can't use ModifyRequest, as the x-goog-api-client header is added later by ConfigurableMessageHandler.
```

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/RetryHandler.cs [hunk 2]
```diff
@@ -65,5 +79,69 @@ public async Task<bool> HandleResponseAsync(HandleUnsuccessfulResponseArgs args)
...
+                if (!request.Headers.TryGetValues(VersionHeaderBuilder.HeaderName, out var values) || values.Count() != 1)
+                {
+                    return Task.CompletedTask;
+                }
+                string value = values.Single();
+                List<string> parts = value.Split(' ').ToList();
+
+                bool gotInvocationId = false;
+                bool gotAttemptCount = false;
+                for (int i = 0; i < parts.Count; i++)
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/RetryHandler.cs:111]: nit:
```suggestion
                for (int i = 0; i < parts.Count && (!gotInvocationid || !gotAttemptCount); i++)
```

**Review by amanda-tarafa — APPROVED** [review]
LGTM!

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/RetryHandler.cs (hunk 1 above)
Comment (jskeet) [reply apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/RetryHandler.cs:55]: Will do.

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/RetryHandler.cs (hunk 2 above)
Comment (jskeet) [reply apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/RetryHandler.cs:111]: Okay, will leave. We're unlikely to add more parts in general - and if we add more parts for retry, this logic would become even hairier (or possibly we'd fail to update it at all...)

**Comment by jskeet** [comment]
Answers to internal questions suggest this is fine - @anshuldavid13 PTAL.

**Review by anshuldavid13 — APPROVED** [review]

**Pull request merged**